### PR TITLE
Upgrade kafka clients to 2.6

### LINKED ISSUE #178: Update kafka-clients to the newer version
`reactor-kafka` depends on `kafka-clients:2.0.0` that is outdated. Having in mind that in many cases `reactor-kafka` is used in conjunction with `spring-boot` it makes sense to update the version. Usually, spring dependencies are managed  by BOM and newer version s forced anyway

```
...
+--- io.projectreactor.kafka:reactor-kafka -> 1.2.2.RELEASE
|    +--- io.projectreactor:reactor-core:3.3.2.RELEASE -> 3.3.6.RELEASE (*)
|    \--- org.apache.kafka:kafka-clients:2.0.0 -> 2.5.0
...
```

`spring-boot` is using the latest version [`kafka-clients:2.6.0`](https://github.com/spring-projects/spring-boot/blob/b69ede4556ec71c05c9498d37312b1f104659ab5/spring-boot-project/spring-boot-dependencies/build.gradle#L894)

### PR DESCRIPTION
**Pull request by amrynsky**
Fixes #178

### CODE REVIEW

**Comment by amrynsky** [comment]
@bsideup please review

**Comment by garyrussell** [comment]
I'll take a look too.

**Comment by garyrussell** [comment]
We can't add lombok in a framework project.

**Comment by bsideup** [comment]
I agree with @garyrussell - @amrynsky could you please reduce the changeset to a bare minimum? Feel free to submit follow up PRs to discuss your other ideas, but this update should be as atomic as possible :)

**Comment by amrynsky** [comment]
> We can't add lombok in a framework project.

the only reason I added lombok is to use in tests https://github.com/reactor/reactor-kafka/pull/180/commits/74fe65f568a1e1c4e6804516cf4a35c3f8b0f398#diff-0bfe3567e83d15b8d3be367d0ebe3f52c291b17ed04831b35df9af8b6e0615f0

Let me delombok it ;-)

**Comment by garyrussell** [comment]
Even with tests only; we can't force people who want to build the project to use lombok.

**Comment by amrynsky** [comment]
@garyrussell removed lombok and rebased

File: build.gradle [hunk 1]
```diff
@@ -134,6 +134,8 @@ configure(allprojects) { project ->
     testCompile "org.testcontainers:kafka:$testcontainersVersion"
     testCompile 'org.awaitility:awaitility:4.0.2'
     testCompile 'org.assertj:assertj-core:3.16.1'
+    // need this dependency because of https://issues.apache.org/jira/browse/KAFKA-10378
```
Comment (bsideup) [inline build.gradle:137]: I _think_ we should wait for 2.6.1, otherwise the users may encounter the same issue

File: src/test/java/reactor/kafka/AbstractKafkaTest.java [hunk 2]
```diff
@@ -54,17 +54,22 @@
...
+import org.testcontainers.utility.DockerImageName;
+import org.testcontainers.utility.TestcontainersConfiguration;
 import reactor.core.publisher.Flux;
 import reactor.kafka.receiver.ReceiverOptions;
 import reactor.kafka.sender.SenderOptions;
 import reactor.kafka.sender.SenderRecord;
 import reactor.kafka.util.TestUtils;
 
 public abstract class AbstractKafkaTest {
+    private static final String KAFKA_DOCKER_TAG = "5.5.1";
```
Comment (bsideup) [inline src/test/java/reactor/kafka/AbstractKafkaTest.java:66]: could you please revert changes in this file? These are unrelated to the Kafka client update

File: src/test/java/reactor/kafka/receiver/KafkaReceiverTest.java [hunk 3]
```diff
@@ -1357,16 +1359,17 @@ private void cancelSubscriptions(boolean failOnError) {
...
-            for (int j = 0; j < partitions; j++) {
-                TopicPartition p = new TopicPartition(topic, j);
-                OffsetAndMetadata offset = c.committed(p);
-                if (offset != null && offset.offset() > 0)
-                    committed += offset.offset();
-            }
-            return committed;
+    private long committedCount(KafkaReceiver<Integer, String> receiver) {
+        return receiver.doOnConsumer(consumer -> {
+            Set<TopicPartition> topicPartitions = IntStream.range(0, partitions)
```
Comment (bsideup) [inline src/test/java/reactor/kafka/receiver/KafkaReceiverTest.java:1364]: is this change really necessary? Did they remove `committed(TopicPartition)`?

File: build.gradle (hunk 1 above)
Comment (amrynsky) [reply build.gradle:137]: this is just test dependency and there s no risk for the framework because test dependencies are not transitive

Comment (amrynsky) [reply build.gradle:137]: as an alternative, we could use `kafka-clients:2.5`. as far as remember it doesn't have this ssue

Comment (bsideup) [reply build.gradle:137]: I know how test dependencies work. But, if you had to add it to make the tests pass, it indicates that someone else out there may also get surprised after updating reactor-kafka and seeing runtime ClassNotFound exceptions

File: src/test/java/reactor/kafka/AbstractKafkaTest.java (hunk 2 above)
Comment (amrynsky) [reply src/test/java/reactor/kafka/AbstractKafkaTest.java:66]: I could revert, but my motivation was to enable testing against the recent version, and `public KafkaContainer()` is deprecated.

Comment (bsideup) [reply src/test/java/reactor/kafka/AbstractKafkaTest.java:66]: When updating dependencies, it is better to not change any unrelated code to avoid incompatibilities.
Feel free to submit a separate PR that polishes the `KafkaContainer` usage (and can be discussed separately) 👍

File: src/test/java/reactor/kafka/receiver/KafkaReceiverTest.java (hunk 3 above)
Comment (amrynsky) [reply src/test/java/reactor/kafka/receiver/KafkaReceiverTest.java:1364]: `OffsetAndMetadata committed(TopicPartition partition)` is deprecated

File: src/test/java/reactor/kafka/AbstractKafkaTest.java (hunk 2 above)
Comment (amrynsky) [reply src/test/java/reactor/kafka/AbstractKafkaTest.java]: extracted into a separate PR https://github.com/reactor/reactor-kafka/pull/182

File: build.gradle (hunk 1 above)
Comment (bsideup) [reply build.gradle:137]: ping

Comment (amrynsky) [reply build.gradle:137]: @bsideup `2.6.1` is not ready yet and it looks like for `2.6.0` this is the only workaround. Should we use `kafka-clients:2.5` instead? 
I understand that we don't want to add 'temporary' dependency but we could clean up when the fix is out and for the time being, we could limit t to `testRuntimeOnly` to avoid any unattended usages.

**Comment by amrynsky** [comment]
finally, `kafka-clients:2.6.1` released with a fix https://issues.apache.org/jira/browse/KAFKA-10378. @bsideup @garyrussell please check

**Review by bsideup — APPROVED** [review]
looks good! Will give @garyrussell a chance to have a look and merge if he also approves it👍

**Comment by garyrussell** [comment]
I will take a look today.

**Comment by garyrussell** [comment]
@bsideup @amrynsky Looks good to me; thanks!

**Comment by amrynsky** [comment]
> @bsideup @amrynsky Looks good to me; thanks!

@garyrussell it looks like PR requires your approval

**Review by garyrussell — APPROVED** [review]

**Comment by bsideup** [comment]
@amrynsky merged! 🎉  Thanks for your contribution 👍
